src/registries/web3/cursors/packages.ts

```diff
@@ -17,94 +17,45 @@ export default class PackagesCursor extends Paged<BN> implements IterableIterato
 
-  private from: string;
+  private registry: Contract;
 
-  private to: string;
+  private packageIds: any;
 
-  constructor(pageSize: BN, length: BN, web3: Web3, from: string, to: string) {
+  constructor(pageSize: BN, length: BN, web3: Web3, registry: Contract, packageIds: any) {
     super(pageSize);
     this.pointer = new BN(0);
     this.length = length.clone();
     this.web3 = web3;
-    this.from = from;
-    this.to = to;
+    this.registry = registry;
+    this.packageIds = packageIds;
+    this.setPages(packageIds);
   }
 
   private getName(): IteratorResult<ResultType> {
     const promise: ResultType = new Promise((resolve) => {
       const packageId = this.getDatum(this.pointer);
       if (packageId === null) {
-        resolve(''); // TODO: empty string or something else?
+        resolve('');
       } else {
-        const data = this.web3.eth.abi.encodeFunctionCall({
-          name: 'getPackageName',
-          type: 'function',
-          inputs: [{
-            type: 'uint',
-            name: 'packageId',
-          }],
-        }, [`0x${packageId.toString('hex')}`]);
-
-        this.web3.eth.call({
-          from: this.from,
-          to: this.to,
-          data,
-        }).then((result) => {
-          resolve(this.web3.eth.abi.decodeParameter('string', result));
-        });
+        this.registry.methods.getPackageName(packageId).call().then((result) => resolve(result));
       }
     });
-
     this.pointer = this.pointer.addn(1);
-
     return {
-      done: true,
+      done: false,
       value: promise,
     };
   }
 
   public next(): IteratorResult<ResultType> {
     if (this.pointer.lt(this.length)) {
-      if (this.hasPage(this.pointer)) {
-        // we have the page, return the number
-        return this.getName();
-      }
-
-      // we don't have the page, get it
-      const offset = this.pointer.sub(this.pointer.mod(this.pageSize));
-      const limit = offset.add(this.pageSize).subn(1);
-
-      const data = this.web3.eth.abi.encodeFunctionCall({
-        name: 'getAllPackageIds',
-        type: 'function',
-        inputs: [{
-          type: 'uint',
-          name: 'offset',
-        }, {
-          type: 'uint',
-          limit: 'limit',
-        }],
-      }, [`0x${offset.toString('hex')}`, `0x${limit.toString('hex')}`]);
-
-      const promise: ResultType = new Promise(() => this.web3.eth.call({
-        from: this.from,
-        to: this.to,
-        data,
-      }).then((result) => {
-        // split packageIds into an array of BNs
-        // set the page
-        // get/resolve the datum
-        const results = this.web3.eth.abi.decodeParameters(['bytes32[]', 'uint'], result);
-        const packageIds = results[0].map((id: string) => new BN(id));
-        this.setPage(this.pointer, packageIds);
-        return this.getName();
-      }));
-
-      return {
-        done: true,
-        value: promise,
-      };
+      return this.getName();
     }
+    return {
+      done: true,
```

Comment:
Why is this `true` now?

Not sure what happened to the pagination here, but I'd been keen to make sure we didn't break intended behavior! Let me know!

Reply:
Yeah - I did some refactoring here, but it's all tested so nothing should be broken. Originally, I couldn't get either of the `Cursor`s to return anything except for empty arrays (see the other comment in this pr). Essentially, i.e. `PackagesCursor` it was generated all the promises to fetch package names before it had retrieved the relevant package ids. So, I split those into two steps - 1. fetch all the package ids (which now happens in the web3 registry service) 2. fetch the corresponding package names via the cursor

> Why is this true now?

Some of the refactoring was based on this, which I saw everywhere
```js
{
  done: false | true // Iteration done or not
  value: T // value of iteration. May not be present when done is true
}
```

So basically, the tests are working, so everything seems to be working fine. But, if I missed something or if there is a better design options -  I'm all ears.

Reply:
Hm, I'll have to circle back here to evaluate the tests, but I guess my first question is: what happens if you have more packages than the page limit in the registry? Do the tests cover that?

I didn't write the original implementation so I'm not sure about what the behavior used to be.

Also note: TypeScript 3.6 introduced much better types for generators/iterators; it's likely worth upgrading to make use of those.

Reply:
Yup, that test case is covered [here](https://github.com/ethpm/ethpm.js/pull/30/files#diff-512d0016943411014c5e07692e3df9ebR50). But, I do plan to extend testing with a much larger registry soon.

good to know about the generators/iterators, I'll look into switching over to use them.